File: src/tools/project.ts
/**
 * Project management tools for KiCAD MCP server
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { CommandFunction, formatKicadResult, makePassthrough } from "./tool-response.js";

export function registerProjectTools(server: McpServer, callKicadScript: CommandFunction) {
  const passthrough = makePassthrough(callKicadScript);
  // Create project tool
  server.tool(
    "create_project",
    "Create a new KiCAD project. Auto-launches the KiCAD UI by default so the IPC backend can attach (realtime sync + transactions); autoLaunch=false skips. Refuses (errorCode PROJECT_EXISTS) if project files already exist unless overwrite=true, or INVALID_NAME if `name` is empty/whitespace. Returns project.path (the .kicad_pro) and boardPath (the .kicad_pcb).",
    {
      path: z
        .string()
        .describe(
          "Directory in which to create <name>.kicad_pro (NOT the .kicad_pro file itself). If a .kicad_pro file path is passed, it is split into directory + name; a name that disagrees with the filename errors PROJECT_NAME_CONFLICT.",
        ),
      name: z
        .string()
        .describe(
          "Project name (without extension), used as the <name>.kicad_pro basename. Must be non-empty (an empty/whitespace name errors INVALID_NAME).",
        ),
      autoLaunch: z
        .boolean()
        .optional()
        .describe("Launch KiCAD UI after creation (default true; false for headless/CI)"),
      overwrite: z
        .boolean()
        .optional()
        .describe("Replace existing .kicad_pro/.kicad_pcb/.kicad_sch (default false: refuse)"),
    },
    passthrough("create_project"),
  );

  // Open project tool
  server.tool(
    "open_project",
    "Open an existing KiCAD project. Accepts `filename` or `path` (interchangeable): a .kicad_pro/.kicad_pcb file, or a directory containing exactly one .kicad_pro (errors NO_PROJECT_IN_DIR / AMBIGUOUS_PROJECT otherwise). A missing file errors FILE_NOT_FOUND, a non-KiCad file UNSUPPORTED_FILE, and a corrupt board PARSE_ERROR — a failed open leaves any previously-loaded project intact. On success returns project.path (the .kicad_pro) and boardPath (the .kicad_pcb). Auto-launches the KiCAD UI by default so the IPC backend can attach (realtime sync + transactions); autoLaunch=false skips.",
    {
      filename: z
        .string()
        .optional()
        .describe("Path to a .kicad_pro or .kicad_pcb file (alias of `path`)"),
      path: z
        .string()
        .optional()
        .describe(
          "Project location: a .kicad_pro/.kicad_pcb file, or the directory containing exactly one .kicad_pro. Provide either this or `filename`.",
        ),
      autoLaunch: z
        .boolean()
        .optional()
        .describe("Launch KiCAD UI after opening (default true; false for headless/CI)"),
    },
    passthrough("open_project"),
  );

  // Save project tool
  server.tool(
    "save_project",
    "Save the currently-loaded KiCAD project (the last one created/opened). The response states which board file was written (savedPath, the .kicad_pcb) and returns project.path (the .kicad_pro) + boardPath (the .kicad_pcb).",
    {
      path: z
        .string()
        .optional()
        .describe(
          "Optional save-as target (a .kicad_pro or .kicad_pcb file). Omit to save the current project in place.",
        ),
      filename: z
        .string()
        .optional()
        .describe("Alias of `path` (legacy) — optional save-as target"),
    },
    passthrough("save_project"),
  );

  // Get project info tool
  server.tool(
    "get_project_info",
    "Get information about the current KiCAD project. Returns project.path (the .kicad_pro) and boardPath (the .kicad_pcb), consistent with create_project / open_project / save_project.",
    {},
    async () => {
      const result = await callKicadScript("get_project_info", {});
      return formatKicadResult(result);
    },
  );

  // Snapshot project tool — copies the project files and, when possible,
  // renders the saved board to a PDF checkpoint.
  server.tool(
    "snapshot_project",
    "Save a named checkpoint snapshot of the project: copies the project files and renders the board to PDF when possible (returned as a `pdf` field; on failure the PDF is omitted and a note explains why), and records the step label. Call after each major step; required by the demo workflow before waiting for user confirmation.",
    {
      step: z.string().describe("Step number or identifier, e.g. '1'"),
      label: z.string().describe("Short checkpoint label, e.g. 'schematic_ok'"),
      prompt: z
        .string()
        .optional()
        .describe("Prompt text saved as PROMPT_step{step}_{timestamp}.md alongside the snapshot"),
    },
    passthrough("snapshot_project"),
  );
}

File: python/schemas/categories/project.py
"""Project tool schema definitions for KiCAD MCP Server.

Following the MCP 2025-06-18 specification for tool definitions.
Split out of the former monolithic schemas/tool_schemas.py.
"""

PROJECT_TOOLS = [
    {
        "name": "create_project",
        "title": "Create New KiCAD Project",
        "description": "Creates a new KiCAD project with PCB board file and optional project configuration. Automatically creates project directory and initializes board with default settings.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "projectName": {
                    "type": "string",
                    "description": "Name of the project (used for file naming)",
                    "minLength": 1,
                },
                "path": {
                    "type": "string",
                    "description": "Directory path where project will be created (defaults to current working directory)",
                },
                "template": {
                    "type": "string",
                    "description": "Optional path to template board file to copy settings from",
                },
                "overwrite": {
                    "type": "boolean",
                    "description": "Replace an existing project. Defaults to false: if the target .kicad_pro/.kicad_pcb/.kicad_sch already exist the tool refuses (errorCode PROJECT_EXISTS) instead of clobbering them.",
                },
            },
            "required": ["projectName"],
        },
    },
    {
        "name": "open_project",
        "title": "Open Existing KiCAD Project",
        "description": "Opens an existing KiCAD project file (.kicad_pro or .kicad_pcb) and loads the board into memory for manipulation.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "filename": {
                    "type": "string",
                    "description": "Path to .kicad_pro or .kicad_pcb file",
                }
            },
            "required": ["filename"],
        },
    },
    {
        "name": "save_project",
        "title": "Save Current Project",
        "description": "Saves the current board to disk. Can optionally save to a new location.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "filename": {
                    "type": "string",
                    "description": "Optional new path to save the board (if not provided, saves to current location)",
                }
            },
        },
    },
    {
        "name": "snapshot_project",
        "title": "Snapshot Project (Checkpoint)",
        "description": "Copies the entire project folder to a new timestamped snapshot directory so you can resume from this checkpoint later without redoing earlier steps. Call this after every successfully completed design step (e.g. after Step 1 schematic, after Step 2 PCB layout) before asking for user confirmation to proceed.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "step": {
                    "type": "string",
                    "description": "Step number or name to include in snapshot folder name, e.g. '1' or '2'",
                },
                "label": {
                    "type": "string",
                    "description": "Optional short label, e.g. 'schematic_ok' or 'layout_ok'",
                },
                "projectPath": {
                    "type": "string",
                    "description": "Project directory path. Auto-detected from loaded board if omitted.",
                },
            },
        },
    },
    {
        "name": "get_project_info",
        "title": "Get Project Information",
        "description": "Retrieves metadata and properties of the currently open project including name, paths, and board status.",
        "inputSchema": {"type": "object", "properties": {}},
    },
]
